polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala

```diff
@@ -225,16 +349,208 @@ class FileBasedRepository(
 ) extends NotebookRepository {
   protected def pathOf(relativePath: String): Path = path.resolve(relativePath)
 
+  private final class FileNotebookRef private (
+    current: Ref[(Int, Notebook)],
+    pending: Queue[Take[Nothing, (NotebookUpdate, Option[Promise[Nothing, (Int, Notebook)]])]],
+    closed: Promise[Throwable, Unit],
+    log: Logging.Service,
+    renameLock: Semaphore,
+    process: Promise[Nothing, Fiber[Nothing, Unit]],
+    saveIntervalMillis: Long = 5000L,
+    maxSaveFails: Int = 12
+  ) extends NotebookRef {
+
+    // TODO: this would be a good place to implement WAL of updates
```

Comment:
Now that we have access to every NotebookUpdate here, we could implement a WAL file that tracks all changes and allows recovery back to any previous state (as long as the WAL is still around)